fre/make/gfdlfremake/makefilefre.py

```diff
@@ -54,97 +58,115 @@ def linklineBuild(self):
-     with open(self.filePath+"/Makefile","a") as fh:
-       for (c,r,o) in sd:
-            libstring = " "
-            for lib in r:
-                 libstring = libstring+lib+"/lib"+lib+".a "
-            cstring = c+"/lib"+c+".a: "
-            fh.write(cstring+libstring+" FORCE"+"\n")
-            if o == "":
-                 fh.write("\t$(MAKE) SRCROOT=$(SRCROOT) BUILDROOT=$(BUILDROOT) MK_TEMPLATE=$(MK_TEMPLATE) --directory="+c+" $(@F)\n")
-            else:
-                 fh.write("\t$(MAKE) SRCROOT=$(SRCROOT) BUILDROOT=$(BUILDROOT) MK_TEMPLATE=$(MK_TEMPLATE) "+o+" --directory="+c+" $(@F)\n")
-       fh.write("FORCE:\n")
-       fh.write("\n")
-# Set up the clean
-       fh.write("clean:\n")
-       for c in self.c:
-            fh.write("\t$(MAKE) --directory="+c+" clean\n")
-# Set up localize
-       fh.write("localize:\n")
-       for c in self.c:
-            fh.write("\t$(MAKE) -f $(BUILDROOT)"+c+" localize\n")
-# Set up distclean
-       fh.write("distclean:\n")
-       for c in self.c:
-            fh.write("\t$(RM) -r "+c+"\n")
-       fh.write("\t$(RM) -r "+self.e+"\n")
-       fh.write("\t$(RM) -r Makefile \n")
+    def __init__(self,exp,libs,srcDir,bldDir,mkTemplatePath):
+        """
+        Brief: Opens Makefile and sets the experiment and other common variables
+        Param:
+            - self The Makefile object
+            - exp Experiment name
+            - libs Additional libraries/linker flags defined by user
+            - srcDir The path to the source directory
+            - bldDir The path to the build directory
+            - mkTemplatePath The path of the template .mk file for compiling
+        """
+        self.e = exp
+        self.l = libs
+        self.src = srcDir
+        self.bld =  bldDir
+        self.template = mkTemplatePath
+        self.c =[] #components
+        self.r=[] #requires
+        self.o=[] #overrides
+        os.system("mkdir -p "+self.bld)
+        self.filePath = self.bld # Needed so that the container and bare metal builds can
+                                 # use the same function to create the Makefile
+
+    def addComponent (self,c,r,o):
+        """
+        Brief: Adds a component and corresponding requires to the list
+        Param: 
+            - self The Makefile object
+            - c The component
+            - r The requires for that componenet
+            - o The overrides for that component
+        """
+        self.c.append(c)
+        self.r.append(r)
+        self.o.append(o)
+
+    def createLibstring (self,c,r,o):
+        """
+        Brief: Sorts the component by how many requires there are for that component
+        Param:
+            - self The Makefile object
+            - c The component
+            - r The requires for that component
+            - o The overrides for that component
+        """
+        d=zip(self.c,self.r,self.o)
+        return(sorted(d,key=lambda values:len(values[1]),reverse=True))
```

Comment:
mmmm i like one-liners but this one is pretty dense. can we break it up just a touch?